java/test/jmri/util/swing/JemmyUtil.java

```diff
@@ -1,120 +1,471 @@
+            try {
+                //note: the JOptionPane.showMessageDialog method splits the message
+                //by the new line character ('\n') and creates JLables for
+                //each line... so that's what we have to match here
+                String[] lines = labelText.split("\\n");
+                for (String line : lines) {
+                    //wait for this label
+                    new JLabelOperator(jfo, line);
+                }
+            } catch (JemmyException ex) {
+                if (debugFlag) {
+                    dumpToXML();
+                    captureScreenshot();
+                }
+                throw ex;
+            }
+        }
+    }
+
+    /**
+     * create a modal dialog operator thread
+     *
+     * @param dialogTitle the title for the dialog
+     * @param messageText the message for the dialog
+     * @param buttonText  the name of the button to press to dismiss
+     * @return the thread that is waiting to dismiss this dialog
+     */
+    public static Thread createModalDialogOperatorThread(String dialogTitle, String messageText, String buttonText) {
+        Thread t = new Thread(() -> {
+            try {
+                // constructor for jdo will wait until the dialog is visible
+                JDialogOperator jdo = new JDialogOperator(dialogTitle);
+                waitForLabels(jdo, messageText);
+                JButtonOperator jbo = new JButtonOperator(jdo, buttonText);
+                jbo.pushNoBlock();
+            } catch (JemmyException ex) {
+                if (debugFlag) {
+                    dumpToXML();
+                    captureScreenshot();
+                }
+                throw ex;
+            }
+        });
+        t.setName(dialogTitle + " '" + messageText + "' Close Dialog Thread");
+        t.start();
+        return t;
+    }
+
+    /*
+ * a component chooser that matches by ToolTipText
+     */
+    public static class ToolTipComponentChooser implements ComponentChooser {
+
+        private String buttonTooltip;
+        private Operator.StringComparator comparator = Operator.getDefaultStringComparator();
+
+        public ToolTipComponentChooser(String buttonTooltip) {
+            this.buttonTooltip = buttonTooltip;
+        }
+
+        public boolean checkComponent(Component comp) {
+            return comparator.equals(((JComponent) comp).getToolTipText(), buttonTooltip);
+        }
+
+        public String getDescription() {
+            return "Component with tooltip \"" + buttonTooltip + "\".";
+        }
+    }
+
+    /**
+     * delay for N milliseconds
+     *
+     * @param milliseconds how log to delay
+     */
+    public static void delay(long milliseconds) {
+        long nextTime = System.currentTimeMillis() + milliseconds;
+        for (long delta = milliseconds; delta > 0;
+                delta = nextTime - System.currentTimeMillis()) {
+            try {
+                Thread.sleep(delta);
```

Comment:
As I said in a previous comment, we really need to rethink this.  I don’t like the idea of having a raw sleep call in any thread during a test.  It only causes the test to take longer.

Reply:
Case in point (sometimes QueueTool().waitEmpty() nor EventTool().waitNoEvent() don't work): In LayoutEditorTests the testToolBarPosition[Left, Bottom, Right, Float] methods all fail AppVeyor if I use the wait methods instead of the delay. (Lower values may work… but I have't tried them. 1/2 second was a WAG.)

Reply:
I am not going to approve this with a raw sleep in place.

Reply:
Good for you. Do you want me to give up on trying to make LE maintainable and improving code coverage? I've spent the better part of December trying to do both with virtually non-existent Jemmy documentation and dealing with all these waitXXX timeouts that should "just work". I agree that NONE of these wait(/delay) methods should be necessary… but they are. Would you feel better if the code limited the sleep delay to 1/4 second? Or smaller?

Reply:
The place where I found sleep to be necessary was the LayoutEditorTest:testToolBarPosition[Left, Bottom, Right, Float] methods. Feel free to give them a look and suggest what they should be waiting for instead.

Reply:
Thinking about this some more, I realized what the more critical problem I have with this is.  There is already a waitFor method in JUnitUtil that waits for a time period in microsecods.

So remove the delay method because it is duplicative, replace uses of it with JUnitUtil.waitFor(int).

We would still prefer to use one of the waitFor methods with a condition if it is at all possible.

Also, the Jemmy operators do wait for a period of time for  an object to appear.  Delays for GUI operations checked with Jemmy operators really should all be handled by Jemmy itself.

Jemmy has a method to get and set timeouts specified by the Timeoutable interface.  https://javadoc.io/doc/org.netbeans/jemmy/latest/org/netbeans/jemmy/Timeoutable.html

Most, if not all, of the operators implement this interface.  You should probably use that instead of an external delay.

Reply:
> Also, the Jemmy operators do wait for a period of time for an object to appear. Delays for GUI operations checked with Jemmy operators really should all be handled by Jemmy itself.

That 110% great… in theory… In practice the testToolBarPositionXXX tests were failing without the (short) 500 mSec delay before the next (waitFor) operation… that would then timeout after 10 seconds.

At this point I'm content to remove all the delay/sleep code and mark those methods @Ignore and move on… I'll "resolve" this after my next commit.

Reply:
Done! All testToolBarPositionXXX tests now use "new EventTool().waitNoEvent(200);" instead of delay and all delay/sleep code has been removed from JemmyUtil.